### PR TITLE
Implement tools for diffing sets of owners of repo paths; update the regex-based CODEOWNERS matching logic

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
## Overview

This PR implements tools for diffing list of owners for all files in given repository, allowing one to quickly determine how the resolved list of owners for all repository file paths will change if a different CODEOWNERS matcher is used, or the contents of the file are changed.

And as such, this PR contributes to:
- #5135
- #2770

And contributes to providing information necessary to unblock this PR:
- #5088
Per this comment:
- https://github.com/Azure/azure-sdk-tools/pull/5088#pullrequestreview-1241376684

This PR builds upon the logic implemented in:
- #5134

## Changes made

- Added `CodeownersDiffTests` class, a series of tools to be used manually to diff owners of files based on differing sets of settings
- Fixed a bug in `MatchedCodeownersEntry`, i.e. the regex-based CODEOWNERS matcher, where the matcher was "eating up slashes too aggressively" when interpreting `**`, resulting in more matches being made than desired. I added tests to `CodeownersFileTests` to cover these cases.

### CODE REVIEW

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs [hunk 1]
```diff
@@ -158,10 +158,11 @@ private Regex ConvertToRegex(string targetPath, string codeownersPath)
 
             pattern = SetPatternSuffix(targetPath, pattern);
 
-            // Note that the "/**/" case is implicitly covered by "**/".
-            pattern = pattern.Replace($"{DoubleStar}/", "(.*)");
-            // This case is necessary to cover suffix case, e.g. "/foo/bar/**".
-            pattern = pattern.Replace($"/{DoubleStar}", "(.*)");
+            pattern = pattern.Replace($"/{DoubleStar}$", "/(.*)$");
```
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs:161]: I'm not exactly clear on why this change was made.

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs [hunk 2]
```diff
@@ -158,11 +158,13 @@ private Regex ConvertToRegex(string targetPath, string codeownersPath)
...
 
-            // Note that the "/**/" case is implicitly covered by "**/".
-            pattern = pattern.Replace($"{DoubleStar}/", "(.*)");
-            // This case is necessary to cover suffix case, e.g. "/foo/bar/**".
-            pattern = pattern.Replace($"/{DoubleStar}", "(.*)");
-            // This case is necessary to cover inline **, e.g. "/a**b/".
+            // Note we can assume there is "/" after "^" because we normalize
+            // the path by prepending "/" if absent.
+            pattern = pattern.Replace($"^/{DoubleStar}", "^(.*)");
+            pattern = pattern.Replace($"/{DoubleStar}/", "((/.*/)|/)");
```
[the remark this answers is not in the document]
Comment (konrad-jamrozik) [reply tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs:164]: We don't want to make trailing `/` optional. This leads to too much ambiguity and inconsistency between our parser, GitHub parser, and .gitignore, on which the GitHub parser is based. I elaborated on that [here](https://github.com/Azure/azure-sdk-tools/issues/4859#issuecomment-1370360622), but I am happy to provide further context if you wish. It is also listed as required in [our rules](https://github.com/Azure/azure-sdk/blob/main/docs/policies/opensource.md#codeowners).

As to why we need `|/` consider this pattern:
`a/**/b` being matched against `a/b`. Without `|/` it wouldn't match, but it has to.

**Comment by konrad-jamrozik** [comment]
@weshaggard I updated the matcher logic. Please have a look once you have a moment :) Thanks!

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs [hunk 3]
```diff
@@ -148,31 +171,34 @@ private Regex ConvertToRegex(string targetPath, string codeownersPath)
...
             }
 
-            pattern = pattern.Replace("**", DoubleStar);
+            // We replace "/**/", not "**", because we disallow "**" in any other context.
+            // Specifically:
+            // - because we normalize the path to start with "/", any prefix "**/" is
+            // effectively "/**/";
+            // - any suffix "/**" is implicit and equivalent to "/", hence we forbid "/**";
+            // - any inline "**" like "/a**/" or "/**a/" or "/a**b/" would be equivalent to
+            // single star, hence we forbid double star, to avoid confusion.
```
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs:180]: Note to self: actually forbid it.

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs [hunk 4]
```diff
@@ -148,31 +171,34 @@ private Regex ConvertToRegex(string targetPath, string codeownersPath)
                     $"\"{DoubleStar}\" or \"{SingleStar}\"");
             }
 
-            pattern = pattern.Replace("**", DoubleStar);
+            // We replace "/**/", not "**", because we disallow "**" in any other context.
```
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs:174]: Note to self: make a pass over all the comments and PR description to ensure they are up to date. The validation rules have changed, we no longer treat target path `/a` as potentially dir.

We also forbid `/**/` if it is a suffix.

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (konrad-jamrozik) [reply tools/code-owners-parser/CodeOwnersParser/MatchedCodeownersEntry.cs:191]: I synced offline with Wes and pushed a commit with the updated matcher logic.
